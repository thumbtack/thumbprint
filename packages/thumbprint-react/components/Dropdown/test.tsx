import React from 'react';
import { noop } from 'lodash';
import { mount, render } from 'enzyme';
import Dropdown from './index';

test('adds `disabled` attribute', () => {
    const wrapper = mount(<Dropdown isDisabled onChange={noop} value="goose" />);
    expect(wrapper.find('select').prop('disabled')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('adds `id` attribute', () => {
    const wrapper = mount(<Dropdown id="Goose" onChange={noop} value="goose" />);
    expect(wrapper.find('select').prop('id')).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('adds `required` attribute', () => {
    const wrapper = mount(<Dropdown isRequired onChange={noop} value="goose" />);
    expect(wrapper.find('select').prop('required')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
});

test('renders an error state with `hasError`', () => {
    const wrapper = mount(<Dropdown hasError onChange={noop} value="goose" />);
    expect(wrapper).toMatchSnapshot();
});

test('renders options that are passed in as `children`', () => {
    const wrapper = mount(
        <Dropdown onChange={noop} value="goose">
            <option>New York</option>
            <option>California</option>
        </Dropdown>,
    );

    expect(wrapper.find('option')).toHaveLength(2);
    expect(wrapper).toMatchSnapshot();
});

test('uses `value` as the selected option', () => {
    const wrapper = render(
        <Dropdown value="ca" onChange={jest.fn}>
            <option value="ny">New York</option>
            <option value="ca">California</option>
        </Dropdown>,
    );
    expect(wrapper.find('select [selected]').val()).toBe('ca');
    expect(wrapper).toMatchSnapshot();
});

test('calls `onChange` function and supplies new value and event', () => {
    const onChange = jest.fn();

    const wrapper = mount(<Dropdown onChange={onChange} value="goose" />);
    const select = wrapper.find('select');

    const event1 = { target: { value: 'He' } };
    select.simulate('change', event1);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenLastCalledWith(event1.target.value, expect.objectContaining(event1));

    const event2 = { target: { value: 'Hello' } };
    select.simulate('change', event2);
    expect(onChange).toHaveBeenCalledTimes(2);
    expect(onChange).toHaveBeenLastCalledWith(event2.target.value, expect.objectContaining(event2));
});

test('calls `onBlur` function when supplied', () => {
    const onBlur = jest.fn();

    const wrapper = mount(<Dropdown onBlur={onBlur} onChange={jest.fn} value="goose" />);

    wrapper.find('select').simulate('blur');
    expect(onBlur).toHaveBeenCalledTimes(1);
});

test('calls `onFocus` function when supplied', () => {
    const onFocus = jest.fn();

    const wrapper = mount(<Dropdown onFocus={onFocus} onChange={jest.fn} value="goose" />);

    wrapper.find('select').simulate('focus');
    expect(onFocus).toHaveBeenCalledTimes(1);
});

test('calls `onClick` function when clicked on', () => {
    const onClick = jest.fn();

    const wrapper = mount(<Dropdown onClick={onClick} onChange={noop} value="goose" />);

    wrapper.find('select').simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
});

test('adds `dataTest` prop', () => {
    const wrapperA = mount(<Dropdown dataTest="Duck" onChange={noop} value="goose" />);
    expect(wrapperA.find('select').prop('data-test')).toEqual('Duck');
    expect(wrapperA).toMatchSnapshot();
});

test('adds `dataTestId` prop', () => {
    const wrapperA = mount(<Dropdown dataTestId="Duck" onChange={noop} value="goose" />);
    expect(wrapperA.find('select').prop('data-testid')).toEqual('Duck');
    expect(wrapperA).toMatchSnapshot();
});

test('adds `name` HTML attribute', () => {
    const wrapperA = mount(<Dropdown name="duck" onChange={noop} value="goose" />);
    expect(wrapperA.find('select').prop('name')).toEqual('duck');
    expect(wrapperA).toMatchSnapshot();
});
