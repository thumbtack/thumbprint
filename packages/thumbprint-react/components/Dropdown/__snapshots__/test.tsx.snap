// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<Dropdown
  dataTest="Duck"
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      data-test="Duck"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <svg
      className="caret"
      fill="none"
      height="18"
      stroke="#2f3033"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="3"
      viewBox="0 0 24 24"
      width="18"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</Dropdown>
`;

exports[`adds \`disabled\` attribute 1`] = `
<Dropdown
  isDisabled={true}
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDisabled selectSizeLarge"
      disabled={true}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <svg
      className="caret"
      fill="none"
      height="18"
      stroke="#d3d4d5"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="3"
      viewBox="0 0 24 24"
      width="18"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</Dropdown>
`;

exports[`adds \`id\` attribute 1`] = `
<Dropdown
  id="Goose"
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      id="Goose"
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <svg
      className="caret"
      fill="none"
      height="18"
      stroke="#2f3033"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="3"
      viewBox="0 0 24 24"
      width="18"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</Dropdown>
`;

exports[`adds \`name\` HTML attribute 1`] = `
<Dropdown
  name="duck"
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      name="duck"
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <svg
      className="caret"
      fill="none"
      height="18"
      stroke="#2f3033"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="3"
      viewBox="0 0 24 24"
      width="18"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</Dropdown>
`;

exports[`adds \`required\` attribute 1`] = `
<Dropdown
  isRequired={true}
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={true}
      value="goose"
    />
    <svg
      className="caret"
      fill="none"
      height="18"
      stroke="#2f3033"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="3"
      viewBox="0 0 24 24"
      width="18"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</Dropdown>
`;

exports[`renders an error state with \`hasError\` 1`] = `
<Dropdown
  hasError={true}
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateError selectSizeLarge"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    />
    <svg
      className="caret"
      fill="none"
      height="18"
      stroke="#ff5a5f"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="3"
      viewBox="0 0 24 24"
      width="18"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</Dropdown>
`;

exports[`renders options that are passed in as \`children\` 1`] = `
<Dropdown
  onChange={[Function]}
  value="goose"
>
  <div
    className="root"
  >
    <select
      className="select selectStateDefault selectSizeLarge"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      required={false}
      value="goose"
    >
      <option>
        New York
      </option>
      <option>
        California
      </option>
    </select>
    <svg
      className="caret"
      fill="none"
      height="18"
      stroke="#2f3033"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="3"
      viewBox="0 0 24 24"
      width="18"
    >
      <polyline
        points="6 9 12 15 18 9"
      />
    </svg>
  </div>
</Dropdown>
`;

exports[`uses \`value\` as the selected option 1`] = `
<div
  class="root"
>
  <select
    class="select selectStateDefault selectSizeLarge"
  >
    <option
      value="ny"
    >
      New York
    </option>
    <option
      selected=""
      value="ca"
    >
      California
    </option>
  </select>
  <svg
    class="caret"
    fill="none"
    height="18"
    stroke="#2f3033"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="3"
    viewBox="0 0 24 24"
    width="18"
  >
    <polyline
      points="6 9 12 15 18 9"
    />
  </svg>
</div>
`;
